### PR TITLE
Refactor `Etcdcopybackupstask` to eliminate Helm chart

### LINKED ISSUE #552: [Feature] Refactor `EtcdCopyBackupsTask` controller to eliminate Helm chart.
**Feature (What you would like to be added):**

Refactor the`EtcdCopyBackupsTask` controller to formulate/create an `EtcdCopyBackupsTask`job using Golang instead of a Helm chart.

**Motivation (Why is this needed?):**

**Approach/Hint to the implement solution (optional):**

### PR DESCRIPTION
**Pull request by seshachalam-yv**
**How to categorize this PR?**

/area robustness
/kind enhancement

**What this PR does / why we need it**:
This PR refactors the `EtcdCopyBackupsTask` controller to formulate/create an `EtcdCopyBackupsTaskjob` using Golang instead of a Helm chart.

**Which issue(s) this PR fixes**:
Fixes #552 #532

**Special notes for your reviewer**:

**Release note**:

```other operator
A bug that caused control plane migrations to fail for shoots with numeric names and a leading `0` has been fixed by eliminating the `EtcdCopyBackupsTask` helm chart.
```
```other developer
Remove helm charts for EtcdCopyBackupsTask controller in favour of go code
```

### CODE REVIEW

File: controllers/etcdcopybackupstask/chart.go [hunk 1]
```diff
@@ -1,129 +0,0 @@
...
-		"storePrefix":     store.Prefix,
-		"storageProvider": storageProvider,
-	}
-	if strings.EqualFold(string(*store.Provider), druidutils.Local) {
-		mountPath, err := druidutils.GetHostMountPathFromSecretRef(ctx, client, logger, store, namespace)
-		if err != nil {
-			return nil, err
-		}
-		storeValues["storageMountPath"] = mountPath
-	}
```
Comment (shreyas-s-rao) [inline controllers/etcdcopybackupstask/chart.go:121]: This case has been missed in the new code, due to which job spec created for etcdcopybackupstask using `local` provider fails to match the original job spec. Specifically, the path that should have been generated is `/etc/gardener/local-backupbuckets/[secret-name]`, whereas the PR generates only `[secret-name]`.

I would recommend writing a test case to test this, so that we don't face any related regressions in the future.

File: controllers/etcdcopybackupstask/reconciler.go [hunk 2]
```diff
@@ -17,16 +17,19 @@ package etcdcopybackupstask
 import (
```
Comment (shreyas-s-rao) [inline controllers/etcdcopybackupstask/reconciler.go:17]: Please write unit tests for the new functions/methods added in the PR, to ensure we avoid regressions in any later refactorings.

**Review by shreyas-s-rao — CHANGES_REQUESTED** [review]
@seshachalam-yv thanks for well-written PR. I have a few suggestions/comments. Please address them. Also, please update the release notes to also add `Remove helm charts for EtcdCopyBackupsTask controller in favour of go code` as a separate note.

File: api/v1alpha1/types_etcdcopybackupstask.go [hunk 3]
```diff
@@ -93,3 +97,7 @@ type EtcdCopyBackupsTaskStatus struct {
 	// +optional
 	LastError *string `json:"lastError,omitempty"`
 }
+
+func (e *EtcdCopyBackupsTask) GetCopyBackupsJobName() string {
```
Comment (unmarshall) [inline api/v1alpha1/types_etcdcopybackupstask.go:102]:
```suggestion
func (e *EtcdCopyBackupsTask) GetJobName() string {
```
The struct on which you are calling the method already is named `EtcdCopyBackupsTask`, there is no need to repeat that in the method as well.

Comment (unmarshall) [inline api/v1alpha1/types_etcdcopybackupstask.go:102]: Please add a doc string to the exposed method.

File: pkg/utils/image.go [hunk 4]
```diff
@@ -56,3 +56,12 @@ func chooseImage(key string, specImage *string, iv imagevector.ImageVector) (*st
 	}
 	return pointer.String(ivImage[key].String()), nil
 }
+
+// GetBackupRestoreImage returns the image for `backup-restore` from the given image vector.
+func GetBackupRestoreImage(iv imagevector.ImageVector) (string, error) {
```
Comment (unmarshall) [inline pkg/utils/image.go:61]: There is an already existing method `GetEtcdImages` which calls `chooseImage`. If you see its implementation it first chooses the image from `etcd.Spec.Backup.Image` and if that is not there then it will take it from image vector. Instead of introducing a new method which duplicates some code (currently incorrectly), you create individual methods to get etcd and backup-restore images and then call these methods also from `GetEtcdImages`

```golang
func GetEtcdImage(etcd *druidv1alpha1.Etcd, iv imagevector.ImageVector) (string, error) {
  // TODO
}
func GetEtcdBackupRestoreImage(etcd *druidv1alpha1.Etcd, iv imagevector.ImageVector) (string, error) {
  // TODO
}
func GetEtcdImages(etcd *druidv1alpha1.Etcd, iv imagevector.ImageVector) (string, string, error) {
   etcdImg, err := GetEtcdImage(etcd, iv)
   if err != nil {
       return nil, nil, err
   }
   etcdBackupRestoreImg, err := GetEtcdBackupRestoreImage(etcd, iv)
   if err != nil {
       return nil, nil, err
   }
   return etcdImg, etcdBackupRestoreImg, nil
}
```

File: controllers/etcdcopybackupstask/reconciler.go [hunk 5]
```diff
@@ -58,24 +58,18 @@ func NewReconciler(mgr manager.Manager, config *Config) (*Reconciler, error) {
...
-		imageVector:   imageVector,
-		chartBasePath: chartBasePath,
-		chartRenderer: chartRenderer,
-		logger:        log.Log.WithName("etcd-copy-backups-task-controller"),
-	}, nil
+		Client:      mgr.GetClient(),
+		Config:      config,
+		imageVector: imageVector,
+		logger:      log.Log.WithName("etcd-copy-backups-task-controller"),
+	}
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:72]: In `reconcile` method can you change the logger to the following:
```go
logger := r.logger.WithValues("etcdCopyBackupsTask", kutil.ObjectName(task), "operation", "reconcile")
```

File: controllers/etcdcopybackupstask/reconciler.go [hunk 6]
```diff
@@ -140,8 +134,8 @@ func (r *Reconciler) doReconcile(ctx context.Context, task *druidv1alpha1.EtcdCo
 		return status, nil
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:134]: Are the jobs getting removed automatically? I think there was another issue where it was found that the TTL was not set on the jobs that enables auto removal. So imagine a case where the user creates a EtcdCopyBackupsTask. It gets completed and he removes the task but the job is still there (check this). Now create another task with the same name. Since the job with name `<task-name>-worker` already exists then no new Job will be created. Can you check if this is the behavior?

File: controllers/etcdcopybackupstask/reconciler.go [hunk 7]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+	sourceProvider, err := druidutils.StorageProviderFromInfraProvider(sourceStore.Provider)
+	if err != nil {
+		return nil, err
+	}
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      task.GetCopyBackupsJobName(),
+			Namespace: task.Namespace,
+			Annotations: map[string]string{
+				common.GardenerOwnedBy:   kutil.Key(task.Namespace, task.GetCopyBackupsJobName()).String(),
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:302]: using `client.ObjectKey` should have been good enough.

File: controllers/etcdcopybackupstask/reconciler.go [hunk 8]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+
+	if len(task.Spec.SourceStore.Prefix) > 0 {
+		*command = append(*command, "--source-store-prefix="+task.Spec.SourceStore.Prefix)
+	}
+
+	if task.Spec.SourceStore.Container != nil && len(*task.Spec.SourceStore.Container) > 0 {
+		*command = append(*command, "--source-store-container="+*task.Spec.SourceStore.Container)
+	}
+
+	if task.Spec.MaxBackupAge != nil && *task.Spec.MaxBackupAge != 0 {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:479]: there is no check on the value of maxbackupage in the chart. Here you are only setting if the value is not 0. What is consequence if the value is set to 0?

File: controllers/etcdcopybackupstask/reconciler.go [hunk 9]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+		*command = append(*command, "--max-backup-age="+strconv.Itoa(int(*task.Spec.MaxBackupAge)))
+	}
+
+	if task.Spec.MaxBackups != nil && *task.Spec.MaxBackups != 0 {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:483]: In the chart there was no check for MaxBackups = 0. Now you have added this check. What is the consequence if MaxBackups = 0? If this is not allowed then this should be part of the API validation.

File: controllers/etcdcopybackupstask/reconciler.go [hunk 10]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+	case druidutils.S3, druidutils.ABS, druidutils.Swift, druidutils.OCS, druidutils.OSS:
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      getVolumeNamePrefix(volumeMountPrefix) + "etcd-backup",
+			MountPath: "/root/" + getVolumeNamePrefix(volumeMountPrefix) + "etcd-backup/",
+		})
+	}
+	return
+}
+
+func getEnvVarFromValues(name, value string) corev1.EnvVar {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:418]: You are not getting EnvVar from Values and therefore the name of this function is not accurate. Please rename this function to `mapToEnvVar`

File: controllers/etcdcopybackupstask/reconciler.go [hunk 11]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+						{
+							Name:            "copy-backups",
+							Image:           etcdBackupImage,
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Command: []string{
+								"etcdbrctl",
+								"copy",
+								"--snapstore-temp-directory=/var/etcd/data/tmp",
+							},
+							Env:          append(getProviderEnvs(&sourceStore, sourceProvider, "SOURCE_", "source-"), getProviderEnvs(&targetStore, targetProvider, "", "")...),
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:326]: can you create a function `createEnvVars`?

File: controllers/etcdcopybackupstask/reconciler.go [hunk 12]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+							VolumeMounts: append(getVolumeMounts(&sourceStore, sourceProvider, "source-"), getVolumeMounts(&targetStore, targetProvider, "target-")...),
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:327]: can you create a function `createVolumeMounts`?

File: controllers/etcdcopybackupstask/reconciler.go [hunk 13]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+		Name:  name,
+		Value: value,
+	}
+}
+
+func getProviderEnvs(store *druidv1alpha1.StoreSpec, storeProvider, prefix, volumePrefix string) []corev1.EnvVar {
+	env := []corev1.EnvVar{getEnvVarFromValues(prefix+"STORAGE_CONTAINER", *store.Container)}
+	switch storeProvider {
+	case druidutils.S3:
+		env = append(env, getEnvVarFromValues(prefix+"AWS_APPLICATION_CREDENTIALS", "/root/"+volumePrefix+"etcd-backup"))
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:429]: can we define constants for these?

File: controllers/etcdcopybackupstask/reconciler.go [hunk 14]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+
+func getEnvVarFromValues(name, value string) corev1.EnvVar {
+	return corev1.EnvVar{
+		Name:  name,
+		Value: value,
+	}
+}
+
+func getProviderEnvs(store *druidv1alpha1.StoreSpec, storeProvider, prefix, volumePrefix string) []corev1.EnvVar {
+	env := []corev1.EnvVar{getEnvVarFromValues(prefix+"STORAGE_CONTAINER", *store.Container)}
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:426]: Proper way is to ensure that you have a fixed sized slice as you will at most have 2 elements in the slice.
```go
	envVars := make([]corev1.EnvVar, 2)
```

File: pkg/component/etcd/poddisruptionbudget/values.go [hunk 15]
```diff
@@ -17,11 +17,9 @@ package poddisruptionbudget
 import "k8s.io/apimachinery/pkg/types"
 
 const (
-	ownedByAnnotationKey   = "gardener.cloud/owned-by"
-	ownerTypeAnnotationKey = "gardener.cloud/owner-type"
-	instanceKey            = "instance"
-	nameKey                = "name"
-	appKey                 = "app"
+	instanceKey = "instance"
```
Comment (unmarshall) [inline pkg/component/etcd/poddisruptionbudget/values.go:20]: These changes are completely unrelated to etcdcopybackuptask and should not be clubbed with this PR. Please remove this and file a separate PR for this.

File: pkg/component/etcd/poddisruptionbudget/valueshelper.go [hunk 16]
```diff
@@ -33,8 +34,8 @@ func GenerateValues(etcd *druidv1alpha1.Etcd) Values {
 	}
 
 	annotations := map[string]string{
-		ownedByAnnotationKey:   fmt.Sprintf("%s/%s", etcd.Namespace, etcd.Name),
-		ownerTypeAnnotationKey: "etcd",
+		common.GardenerOwnedBy:   fmt.Sprintf("%s/%s", etcd.Namespace, etcd.Name),
```
Comment (unmarshall) [inline pkg/component/etcd/poddisruptionbudget/valueshelper.go:37]: These changes are completely unrelated to etcdcopybackuptask and should not be clubbed with this PR. Please remove this and file a separate PR for this.

File: pkg/component/etcd/statefulset/statefulset.go [hunk 17]
```diff
@@ -465,15 +466,13 @@ func getObjectMeta(val *Values, sts *appsv1.StatefulSet) metav1.ObjectMeta {
...
-	ownedByAnnotationKey          = "gardener.cloud/owned-by"
-	ownerTypeAnnotationKey        = "gardener.cloud/owner-type"
 )
 
 func getStsAnnotations(val *Values, sts *appsv1.StatefulSet) map[string]string {
 	annotations := utils.MergeStringMaps(
 		map[string]string{
-			ownedByAnnotationKey:   fmt.Sprintf("%s/%s", val.Namespace, val.Name),
-			ownerTypeAnnotationKey: "etcd",
+			common.GardenerOwnedBy:   fmt.Sprintf("%s/%s", val.Namespace, val.Name),
```
Comment (unmarshall) [inline pkg/component/etcd/statefulset/statefulset.go:474]: These changes are completely unrelated to etcdcopybackuptask and should not be clubbed with this PR. Please remove this and file a separate PR for this.

File: pkg/component/etcd/statefulset/statefulset_test.go [hunk 18]
```diff
@@ -422,12 +422,12 @@ func checkStatefulset(sts *appsv1.StatefulSet, values Values) {
 			"Name":      Equal(values.Name),
 			"Namespace": Equal(values.Namespace),
 			"Annotations": MatchAllKeys(Keys{
-				"checksum/etcd-configmap":   Equal("abc123"),
-				"gardener.cloud/owned-by":   Equal(fmt.Sprintf("%s/%s", values.Namespace, values.Name)),
-				"gardener.cloud/owner-type": Equal("etcd"),
-				"app":                       Equal("etcd-statefulset"),
-				"role":                      Equal("test"),
-				"instance":                  Equal(values.Name),
+				"checksum/etcd-configmap": Equal("abc123"),
```
Comment (unmarshall) [inline pkg/component/etcd/statefulset/statefulset_test.go:425]: These changes are completely unrelated to etcdcopybackuptask and should not be clubbed with this PR. Please remove this and file a separate PR for this.

File: test/integration/controllers/etcd/reconciler_test.go [hunk 19]
```diff
@@ -605,13 +605,13 @@ func validateDefaultValuesForEtcd(instance *druidv1alpha1.Etcd, s *appsv1.Statef
...
 			"Namespace": Equal(instance.Namespace),
 			"Annotations": MatchAllKeys(Keys{
-				"gardener.cloud/owned-by":   Equal(fmt.Sprintf("%s/%s", instance.Namespace, instance.Name)),
-				"gardener.cloud/owner-type": Equal("etcd"),
-				"app":                       Equal("etcd-statefulset"),
-				"role":                      Equal("test"),
-				"instance":                  Equal(instance.Name),
-				"checksum/etcd-configmap":   Equal(configMapChecksum),
-				"name":                      Equal("etcd"),
+				common.GardenerOwnedBy:    Equal(fmt.Sprintf("%s/%s", instance.Namespace, instance.Name)),
```
Comment (unmarshall) [inline test/integration/controllers/etcd/reconciler_test.go:608]: These changes are completely unrelated to etcdcopybackuptask and should not be clubbed with this PR. Please remove this and file a separate PR for this.

File: controllers/etcdcopybackupstask/reconciler.go (hunk 6 above)
Comment (seshachalam-yv) [reply controllers/etcdcopybackupstask/reconciler.go:140]: I am adding `EtcdCopyBackupsTask ` as ownerreference to job `<task-name>-worker` [here](https://github.com/gardener/etcd-druid/pull/553/files#diff-50d8f23e4c9d06e06ce9338c5d105b870667137a087aa14dc4f0a257da4c7730R372-R374).
This ensures that once the `EtcdCopyBackupsTask` task is deleted, its dependents will also be deleted.

File: controllers/etcdcopybackupstask/reconciler.go [hunk 20]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+					},
+				},
+				Spec: corev1.PodSpec{
+					RestartPolicy: corev1.RestartPolicyOnFailure,
+					Containers: []corev1.Container{
+						{
+							Name:            "copy-backups",
+							Image:           etcdBackupImage,
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Command: []string{
```
[the remark this answers is not in the document]
Comment (seshachalam-yv) [reply controllers/etcdcopybackupstask/reconciler.go]: Fixed

File: controllers/etcdcopybackupstask/reconciler.go [hunk 21]
```diff
@@ -283,6 +277,218 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+		env = append(env, getEnvVarFromValues(prefix+"OPENSTACK_APPLICATION_CREDENTIALS", "/root/"+volumePrefix+"etcd-backup"))
+	case druidutils.OCS:
+		env = append(env, getEnvVarFromValues(prefix+"OPENSHIFT_APPLICATION_CREDENTIALS", "/root/"+volumePrefix+"etcd-backup"))
+	case druidutils.OSS:
+		env = append(env, getEnvVarFromValues(prefix+"ALICLOUD_APPLICATION_CREDENTIALS", "/root/"+volumePrefix+"etcd-backup"))
+	}
+	return env
+}
+
+func setEtcdBackupJobCommands(task *druidv1alpha1.EtcdCopyBackupsTask, command *[]string) error {
```
[the remark this answers is not in the document]
Comment (seshachalam-yv) [reply controllers/etcdcopybackupstask/reconciler.go]: Yeah, Madhav. Now I have introduced two functions, `createJobCommandFromStore`
`createJobCommandFromTask` to construct the command by reusing the source/target store..

File: controllers/etcdcopybackupstask/reconciler.go (hunk 9 above)
Comment (seshachalam-yv) [reply controllers/etcdcopybackupstask/reconciler.go]: Fixed

**Comment by seshachalam-yv** [comment]
Thank you @shreyas-s-rao, @unmarshall, and @aaronfern for taking the time to review my PR. 
I have addressed your comments and made the necessary changes. 

**One pending item is adding unit tests for the new functions that I introduced. I will add the tests as soon as you have reviewed the updated code.**

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 22]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+				errors := testutils.CreateSecrets(ctx, fakeClient, task.Namespace, task.Spec.SourceStore.SecretRef.Name, task.Spec.TargetStore.SecretRef.Name)
+				Expect(errors).Should(BeNil())
+
+				job, err := reconciler.createJobObject(ctx, task)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(job).Should(PointTo(matchJob(task, reconciler.imageVector)))
+			},
+			Entry("with #Local provider",
+				"foo01", druidv1alpha1.StorageProvider("Local"), true),
+			Entry("should create the job, update the task status, and delete the job if the job failed",
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:176]: The description of this entry is not clear. If you see the rest of the entries then it seems what you are testing is a creation of the Job object and then checking if the Job created is as expected. However this entry says that it should also delete the job if it failed but i do not see any check for that in the function

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 23]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+			It("should return error", func() {
+				task := testutils.CreateEtcdCopyBackupsTask("test", namespace, "Local", true)
+				task.Spec.SourceStore.Provider = &unknownProvider
+				job, err := reconciler.createJobObject(ctx, task)
+
+				Expect(job).To(BeNil())
+				Expect(err).To(HaveOccurred())
+			})
+		})
+		Context("when createVolumesFromstore function fails while creating volumes from the source store", func() {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:222]: Apart from the description what is different between `when createVolumesFromstore function fails while creating volumes from the source store` and `when source store provider is unknown`. The body of `It` is identical.

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 24]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+				task := testutils.CreateEtcdCopyBackupsTask("test", namespace, "Local", true)
+				task.Spec.SourceStore.Provider = &unknownProvider
+				job, err := reconciler.createJobObject(ctx, task)
+
+				Expect(job).To(BeNil())
+				Expect(err).To(HaveOccurred())
+			})
+		})
+
+		Context("when createVolumesFromstore function fails while creating volumes from the target store", func() {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:233]: Same as above

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 25]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+			It("returns a command slice with provider, prefix, and container information", func() {
+				expected := []string{
+					"--prefixstorage-provider=storage_provider",
+					"--prefixstore-prefix=store_prefix",
+					"--prefixstore-container=store_container",
+				}
+				Expect(createJobCommandFromStore(&store, provider, prefix)).To(Equal(expected))
+			})
+
+			It("returns a command slice with provider and prefix information only", func() {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:277]:
```suggestion
			It("when StoreSpec.Container is nil, should return a command slice with provider and prefix information only", func() {
```

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 26]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+				expected := []string{
+					"--prefixstorage-provider=storage_provider",
+					"--prefixstore-prefix=store_prefix",
+				}
+				store.Container = nil
+				Expect(createJobCommandFromStore(&store, provider, prefix)).To(Equal(expected))
+			})
+
+			It("returns a command slice with provider and container information only", func() {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:286]:
```suggestion
			It("when StoreSpec.Prefix is empty, should return a command slice with provider and container information only", func() {
```

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 27]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+				expected := []string{
+					"--prefixstorage-provider=storage_provider",
+					"--prefixstore-container=store_container",
+				}
+				store.Prefix = ""
+				Expect(createJobCommandFromStore(&store, provider, prefix)).To(Equal(expected))
+			})
+
+			It("returns a command slice with prefix and container information only", func() {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:295]:
```suggestion
			It("when StoreSpec.Provider is nil, returns a command slice with prefix and container information only", func() {
```

Comment (unmarshall) [reply controllers/etcdcopybackupstask/reconciler_test.go:295]: Should it be allowed that a `StoreSpec` is specified but the provider is nil?

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 28]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+						Fail(fmt.Sprintf("Unknown provider: %s", provider))
+					}
+
+					Expect(volumeMounts[0].Name).To(Equal(expectedMountName))
+					Expect(volumeMounts[0].MountPath).To(Equal(expectedMountPath))
+				})
+			})
+		}
+
+		Context("with provider #Local", func() {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:481]: you have already included provider local in the test case above. Any reason to create another test again for local provider?

File: controllers/etcdcopybackupstask/reconciler.go [hunk 29]
```diff
@@ -283,6 +283,238 @@ func getConditionType(jobConditionType batchv1.JobConditionType) druidv1alpha1.C
...
+	if prefix == sourcePrefix {
+		return prefix
+	}
+	return ""
+}
+
+// createVolumesFromstore generates a slice of VolumeMounts for an EtcdCopyBackups job based on the given StoreSpec and
+// provider. The prefix is used to differentiate between source and target volume.
+// This function creates the necessary Volume configurations for various storage providers.
+func (r *Reconciler) createVolumesFromstore(ctx context.Context, store *druidv1alpha1.StoreSpec, namespace, provider, prefix string) (volumes []corev1.Volume, err error) {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler.go:409]:
```suggestion
func (r *Reconciler) createVolumesFromStore(ctx context.Context, store *druidv1alpha1.StoreSpec, namespace, provider, prefix string) (volumes []corev1.Volume, err error) {
```

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 30]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+			It("should create the correct volume mounts", func() {
+				volumeMounts := createVolumeMountsFromStore(storeSpec, druidutils.Local, volumeMountPrefix)
+				Expect(volumeMounts).To(HaveLen(1))
+				Expect(volumeMounts[0].Name).To(Equal(volumeMountPrefix + "host-storage"))
+				Expect(volumeMounts[0].MountPath).To(Equal(*storeSpec.Container))
+			})
+		})
+	})
+
+	Describe("#createVolumesFromstore", func() {
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:499]:
```suggestion
	Describe("#createVolumesFromStore", func() {
```

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 31]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+			})
+
+			It("should create the correct volumes when secret data hostPath is set", func() {
+				secret.Data = map[string][]byte{
+					druidutils.EtcdBackupSecretHostPath: []byte("/test/hostPath"),
+				}
+				Expect(fakeClient.Create(ctx, secret)).To(Succeed())
+
+				store.SecretRef = &corev1.SecretReference{Name: secret.Name}
+				volumes, err := reconciler.createVolumesFromstore(ctx, store, namespace, string(providerLocal), "source-")
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:539]:
```suggestion
				volumes, err := reconciler.createVolumesFromStore(ctx, store, namespace, string(providerLocal), "source-")
```

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 32]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+				Expect(hostPathVolumeSource).NotTo(BeNil())
+				Expect(hostPathVolumeSource.Path).To(Equal("/test/hostPath/" + *store.Container))
+				Expect(*hostPathVolumeSource.Type).To(Equal(corev1.HostPathDirectory))
+			})
+
+			It("should create the correct volumes when secret data hostPath is not set", func() {
+				Expect(fakeClient.Create(ctx, secret)).To(Succeed())
+
+				store.SecretRef = &corev1.SecretReference{Name: secret.Name}
+				volumes, err := reconciler.createVolumesFromstore(ctx, store, namespace, string(providerLocal), "source-")
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:554]:
```suggestion
				volumes, err := reconciler.createVolumesFromStore(ctx, store, namespace, string(providerLocal), "source-")
```

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 33]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+				Expect(hostPathVolumeSource).NotTo(BeNil())
+				Expect(hostPathVolumeSource.Path).To(Equal(druidutils.LocalProviderDefaultMountPath + "/" + *store.Container))
+				Expect(*hostPathVolumeSource.Type).To(Equal(corev1.HostPathDirectory))
+			})
+
+			It("should create the correct volumes when store.SecretRef is not refered", func() {
+				Expect(fakeClient.Create(ctx, secret)).To(Succeed())
+
+				store.SecretRef = &corev1.SecretReference{Name: secret.Name}
+				volumes, err := reconciler.createVolumesFromstore(ctx, store, namespace, string(providerLocal), "source-")
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:569]:
```suggestion
				volumes, err := reconciler.createVolumesFromStore(ctx, store, namespace, string(providerLocal), "source-")
```

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 34]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+
+					AfterEach(func() {
+						// Clean up secret after each test case
+						Expect(fakeClient.Delete(ctx, secret)).To(Succeed())
+					})
+
+					It("should create the correct volumes", func() {
+						// Call the function being tested with a valid secret reference
+						store.SecretRef = &corev1.SecretReference{Name: secret.Name}
+						volumes, err := reconciler.createVolumesFromstore(ctx, store, namespace, string(storageProvider), "source-")
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:630]:
```suggestion
						volumes, err := reconciler.createVolumesFromStore(ctx, store, namespace, string(storageProvider), "source-")
```

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 35]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+						Expect(volumeSource).NotTo(BeNil())
+						Expect(volumeSource.Secret).NotTo(BeNil())
+						Expect(*volumeSource.Secret).To(Equal(corev1.SecretVolumeSource{
+							SecretName: store.SecretRef.Name,
+						}))
+					})
+
+					It("should return an error when secret reference is invalid", func() {
+						// Call the function being tested with an invalid secret reference
+						volumes, err := reconciler.createVolumesFromstore(ctx, store, namespace, string(storageProvider), "source-")
```
Comment (unmarshall) [inline controllers/etcdcopybackupstask/reconciler_test.go:646]:
```suggestion
						volumes, err := reconciler.createVolumesFromStore(ctx, store, namespace, string(storageProvider), "source-")
```

File: controllers/etcdcopybackupstask/reconciler_test.go (hunk 22 above)
Comment (seshachalam-yv) [reply controllers/etcdcopybackupstask/reconciler_test.go]: I was trying to add test case with and without optional fields.
Now fixed test case msg properly.
Thanks for catching it

File: controllers/etcdcopybackupstask/reconciler_test.go [hunk 36]
```diff
@@ -132,10 +139,525 @@ var _ = Describe("EtcdCopyBackupsTaskController", func() {
...
+					Name:       common.BackupRestore,
+					Repository: "test-repo",
+					Tag:        pointer.String("etcd-test-tag"),
+				}},
+			}
+		})
+
+		DescribeTable("should create the expected job object with correct metadata, pod template, and containers for a valid input task",
+			func(taskName string, provider druidv1alpha1.StorageProvider, withOptionalFields bool) {
+				task := testutils.CreateEtcdCopyBackupsTask(taskName, namespace, provider, true)
```
[the remark this answers is not in the document]
Comment (seshachalam-yv) [reply controllers/etcdcopybackupstask/reconciler_test.go]: Fixed, now we are passing argument not static value.

File: controllers/etcdcopybackupstask/reconciler_test.go (hunk 23 above)
Comment (seshachalam-yv) [reply controllers/etcdcopybackupstask/reconciler_test.go]: Yes, you are right.
My bad it is duplicate test and I forgot to remove it.
Fixed 
Thanks

File: controllers/etcdcopybackupstask/reconciler_test.go (hunk 24 above)
Comment (seshachalam-yv) [reply controllers/etcdcopybackupstask/reconciler_test.go]: Yes, you are right.
My bad it is duplicate test and I forgot to remove it.
Fixed 
Thanks

**Review by unmarshall — APPROVED** [review]
/lgtm
